In [33]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [34]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
# Read the CSV file into a Pandas DataFrame
df = pd.read_csv(
    Path('Resources/lending_data.csv')   
)

# Review the DataFrame
df.head()


loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [35]:
# Separate the data into labels and features

y = df['loan_status']

X = df.drop(columns=['loan_status'])


In [36]:
# Review the y variable Series

y.head()


0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [37]:
# Review the X variable DataFrame
X.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [38]:
# Check the balance of our target values
target_balance = y.value_counts()

print(target_balance)

0    75036
1     2500
Name: loan_status, dtype: int64


### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [39]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [40]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression


logistic_regression_model = LogisticRegression(random_state=1)

# Fit and save the logistic regression model using the training data
lr_model = logistic_regression_model.fit(X_train, y_train)


### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [41]:
# Make a prediction using the testing data
#Generate training predictions
training_predictions = lr_model.predict(X_train)

#Generate testing predictions
testing_predictions = logistic_regression_model.predict(X_test)


### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [42]:
# Print the balanced_accuracy score of the model
# Import the model for sklearn confusion matrix
from sklearn.metrics import confusion_matrix


In [43]:
# Generate a confusion matrix for the model
training_matrix = confusion_matrix(y_train, training_predictions)

In [44]:
# Print the classification report for the model
print(training_matrix)

[[55994   277]
 [  181  1700]]


### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The logistic regression model performs well in predicting both healthy loans (label 0) and high-risk loans (label 1), but there are slight areas for improvement:

For healthy loans (label 0), the model has high precision and recall but could reduce false positives (instances incorrectly classified as healthy loans).

For high-risk loans (label 1), the model also has good precision and recall, indicating effective identification.

Overall, the model demonstrates a solid performance in predicting both labels combined, as indicated by accuracy.

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [45]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model with a random_state parameter of 1
ros = RandomOverSampler(random_state=1)

# Fit the original training data to the random oversampler model
X_resampled, y_resampled = ros.fit_resample(X_train, y_train)

In [46]:
# Count the distinct values of the resampled labels data

unique_labels = np.unique(y_resampled)

# Print the count of distinct values
print("Distinct Values in Resampled Labels:", len(unique_labels))

Distinct Values in Resampled Labels: 2


### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [47]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
model = LogisticRegression(random_state=1)

# Fit the model using the resampled training data
model.fit(X_resampled, y_resampled)

# Make predictions using the testing data
testing_predictions = model.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [48]:
# Print the balanced_accuracy score of the model 
# Calculate the balanced accuracy score for testing predictions
balanced_accuracy = balanced_accuracy_score(y_test, testing_predictions)

# Print the balanced accuracy score
print("Balanced Accuracy Score:", balanced_accuracy)

Balanced Accuracy Score: 0.9936781215845847


In [49]:
# Generate a confusion matrix for the model
# Generate a confusion matrix for the testing data
confusion_matrix_test = confusion_matrix(y_test, testing_predictions)

# Print the confusion matrix
print("Confusion Matrix for Testing Data:")
print(confusion_matrix_test)

Confusion Matrix for Testing Data:
[[18649   116]
 [    4   615]]


In [50]:
# Print the classification report for the model
classification_report_test = classification_report(y_test, testing_predictions)

# Print the classification report
print("Classification Report for Testing Data:")
print(classification_report_test)

Classification Report for Testing Data:
              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.84      0.99      0.91       619

    accuracy                           0.99     19384
   macro avg       0.92      0.99      0.95     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** This indicates that the model is extremely accurate in identifying healthy loans. It correctly identifies 99% of actual healthy loans (high recall), and when it predicts a loan as healthy, it's almost always correct (high precision).
The model also performs well in identifying high-risk loans. While the precision is slightly lower compared to the recall, it still maintains a good balance between precision and recall with an F1-score of 0.91. This means that it correctly identifies the majority of high-risk loans while maintaining a reasonable level of precision.